nvflare/app_common/ccwf/client_ctl.py

```diff
@@ -0,0 +1,631 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import copy
+import threading
+import time
+from abc import abstractmethod
+
+from nvflare.apis.event_type import EventType
+from nvflare.apis.executor import Executor
+from nvflare.apis.fl_constant import FLContextKey, ReturnCode
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.impl.client_controller import ClientController, Task
+from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.signal import Signal
+from nvflare.app_common.abstract.learnable import Learnable
+from nvflare.app_common.abstract.learnable_persistor import LearnablePersistor
+from nvflare.app_common.abstract.shareable_generator import ShareableGenerator
+from nvflare.app_common.app_constant import AppConstants
+from nvflare.app_common.app_event_type import AppEventType
+from nvflare.app_common.ccwf.common import Constant, ResultType, StatusReport, make_task_name, topic_for_end_workflow
+from nvflare.fuel.utils.validation_utils import check_positive_number, check_str
+from nvflare.security.logging import secure_format_traceback
+
+
+class _LearnTask:
+    def __init__(self, task_name: str, task_data: Shareable, fl_ctx: FLContext):
+        self.task_name = task_name
+        self.task_data = task_data
+        self.fl_ctx = fl_ctx
+        self.abort_signal = Signal()
+
+
+class ClientSideController(Executor, ClientController):
```

Comment:
I have the same feeling but not sure what would be the best names.
Maybe we should rename ClientController to something like TaskController, because it is actually not specific to client side. In the future, we may even rewrite the server-side controller to be based on TaskController.

Reply:
renamed to TaskController.